main.go

```diff
@@ -199,30 +168,33 @@ func main() {
 		})
 	}
 	{
-		var cancel func()
-
-		if opts.Duration != 0 {
-			ctx, cancel = context.WithTimeout(ctx, opts.Duration)
-		} else {
-			ctx, cancel = context.WithCancel(ctx)
-		}
-
-		t := time.NewTicker(opts.Period)
+		if opts.AddressRead != nil {
+			logger := log.With(logger, "component", "querier")
+			t := time.NewTicker(opts.Period)
+
+			var cancel func()
+			if opts.Duration != 0 {
+				ctx, cancel = context.WithTimeout(ctx, opts.Duration+opts.Period)
+			} else {
+				ctx, cancel = context.WithCancel(ctx)
+			}
 
-		if opts.EndpointRead != "" {
 			g.Add(func() error {
+				level.Info(logger).Log("msg", "waiting for a single period before querying for metrics")
+				time.Sleep(opts.InitialQueryWait)
+
 				level.Info(logger).Log("msg", "start querying for metrics")
 				for {
 					select {
 					case <-t.C:
-						if err := read(ctx, endpointRead, opts.Labels, opts.Latency); err != nil {
-							queryResponses.WithLabelValues("error").Inc()
+						if err := read(ctx, opts.AddressRead, opts.Labels, -1*(opts.InitialQueryWait-opts.Period), opts.Latency, m); err != nil {
```

Comment:
This parameter still might need tweaking.

Reply:
Maybe `-1*opts.InitialQueryWait` would better, also I can rename the arg as `shift interval`.

Reply:
I think initial delay is much clearer than shift delay. However, I wouldn’t over engineer this logic. Rather keep it simple, otherwise we are introducing lots of untested complexity for not very large gains imo